contracts/utils/SwapAdapter.sol

```diff
@@ -0,0 +1,134 @@
+
+abstract contract SwapAdapter {
+  using SafeERC20 for IERC20;
+  using Address for address;
+  using Address for address payable;
+
+  /// @notice Describes how the allowance should be revoked for the given spender
+  struct RevokeAction {
+    address spender;
+    IERC20[] tokens;
+  }
+
+  /// @notice Thrown when one of the parameters is a zero address
+  error ZeroAddress();
+
+  /// @notice Thrown when the allowance target is not allowed by the swapper registry
+  error InvalidAllowanceTarget(address spender);
+
+  address public constant PROTOCOL_TOKEN = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+
+  ISwapperRegistry public immutable SWAPPER_REGISTRY;
+
+  constructor(address _swapperRegistry) {
+    if (_swapperRegistry == address(0)) revert ZeroAddress();
+    SWAPPER_REGISTRY = ISwapperRegistry(_swapperRegistry);
+  }
+
+  /**
+   * @notice Takes the given amount of tokens from the caller
+   * @param _token The token to check
+   * @param _amount The amount to take
+   */
+  function _takeFromMsgSender(IERC20 _token, uint256 _amount) internal virtual {
+    _token.safeTransferFrom(msg.sender, address(this), _amount);
+  }
+
+  /**
+   * @notice Executes a swap for the given swapper
+   * @param _swapper The actual swapper
+   * @param _swapData The swap execution data
+   */
+  function _executeSwap(
+    address _swapper,
+    bytes calldata _swapData,
+    uint256 _value
+  ) internal virtual {
+    _swapper.functionCallWithValue(_swapData, _value);
+  }
+
+  /**
+   * @notice Transfers the given amount of tokens from the contract to the recipient
+   * @param _token The token to check
+   * @param _amount The amount to send
+   * @param _recipient The recipient
+   */
+  function _sendToRecipient(
+    address _token,
+    uint256 _amount,
+    address _recipient
+  ) internal virtual {
+    if (_recipient == address(0)) _recipient = msg.sender;
+    if (_token == PROTOCOL_TOKEN) {
+      payable(_recipient).sendValue(_amount);
+    } else {
+      IERC20(_token).safeTransfer(_recipient, _amount);
+    }
+  }
+
+  /**
+   * @notice Checks if the contract has any balance of the given token, and if it does,
+   *         it sends it to the given recipient
+   * @param _token The token to check
+   * @param _recipient The recipient of the token balance
+   */
+  function _sendBalanceOnContractToRecipient(address _token, address _recipient) internal virtual {
+    uint256 _balance = _token == PROTOCOL_TOKEN ? address(this).balance : IERC20(_token).balanceOf(address(this));
+    if (_balance > 0) {
+      _sendToRecipient(_token, _balance, _recipient);
+    }
+  }
+
+  /**
+   * @notice Checks if the given spender has enough allowance, and approves the max amount
+   *         if it doesn't
+   * @param _token The token to check
+   * @param _spender The spender to check
+   * @param _minAllowance The min allowance. If the spender has over this amount, then no extra approve is needed
+   */
+  function _maxApproveSpenderIfNeeded(
```

Comment:
Let's remove this function and simply use OpenZeppelin's `forceApprove(spender, max(uint256))`. We might need to update to the latest version of the library